File: rust/tests/catalogue_conformance_test.rs
#![allow(clippy::unwrap_used)]

use github_copilot_sdk::rpc::{CatalogCandidate, CatalogCandidateSource, CatalogSearchResult};

const OPAQUE_MCP_HANDLE: &str = "opaque:mcp/01-do-not-parse";
const OPAQUE_SKILL_HANDLE: &str = "opaque:skill/02-do-not-parse";

#[test]
fn catalog_search_result_preserves_candidate_semantics() {
    let result: CatalogSearchResult = serde_json::from_value(serde_json::json!({
        "kind": "succeeded",
        "searchId": "search-01",
        "candidates": [
            {
                "kind": "mcp-server",
                "handle": OPAQUE_MCP_HANDLE,
                "handleExpiresAt": "2026-09-02T12:00:00Z",
                "mediaType": "application/mcp-server-card+json",
                "installability": "installable",
                "displayName": "Example MCP",
                "rawCard": {"secret": "must-not-survive"},
                "source": {"kind": "url", "url": "https://catalog.example/mcp.json"},
                "provenance": {
                    "authority": "catalog.example",
                    "observedAt": "2026-09-02T11:00:00Z",
                    "mediaType": "application/mcp-server-card+json"
                }
            },
            {
                "kind": "ai-skill",
                "handle": OPAQUE_SKILL_HANDLE,
                "handleExpiresAt": "2026-09-02T12:00:00Z",
                "mediaType": "application/ai-skill",
                "installability": "not-installable-kind",
                "displayName": "Example skill",
                "rawCard": {"secret": "must-not-survive"},
                "source": {"kind": "embedded"},
                "provenance": {
                    "authority": "catalog.example",
                    "observedAt": "2026-09-02T11:00:00Z",
                    "mediaType": "application/ai-skill"
                }
            }
        ],
        "truncated": false,
        "negotiated": {
            "runtimeProtocolVersion": 1,
            "grantedCapabilities": ["mcp-server-card", "ai-skill-discovery"]
        }
    }))
    .unwrap();

    let CatalogSearchResult::Succeeded(success) = &result else {
        panic!("expected a successful catalogue search");
    };
    let CatalogCandidate::McpServer(mcp) = &success.candidates[0] else {
        panic!("expected an MCP server candidate");
    };
    let CatalogCandidate::AiSkill(skill) = &success.candidates[1] else {
        panic!("expected an AI skill candidate");
    };
    assert_eq!(mcp.handle, OPAQUE_MCP_HANDLE);
    assert_eq!(skill.handle, OPAQUE_SKILL_HANDLE);
    assert!(matches!(mcp.source, CatalogCandidateSource::Url(_)));
    assert!(matches!(skill.source, CatalogCandidateSource::Embedded(_)));

    let wire = serde_json::to_value(&result).unwrap();
    for candidate in wire["candidates"].as_array().unwrap() {
        let fields = candidate.as_object().unwrap();
        for forbidden in ["card", "cardData", "rawCard"] {
            assert!(
                !fields.contains_key(forbidden),
                "candidate leaked {forbidden}"
            );
        }
    }
}

#[test]
fn catalog_search_result_preserves_refusals_and_failures() {
    let authentication: CatalogSearchResult = serde_json::from_value(serde_json::json!({
        "kind": "authentication-required",
        "reason": "no-credential",
        "message": "Sign in is required."
    }))
    .unwrap();
    assert!(matches!(
        authentication,
        CatalogSearchResult::AuthenticationRequired(_)
    ));

    let network: CatalogSearchResult = serde_json::from_value(serde_json::json!({
        "kind": "network-failure",
        "reason": "timeout",
        "retryAfterSeconds": 30,
        "message": "The catalogue timed out."
    }))
    .unwrap();
    let CatalogSearchResult::NetworkFailure(failure) = network else {
        panic!("expected a network failure");
    };
    assert_eq!(failure.retry_after_seconds, Some(30));
}

File: go/rpc/catalogue_conformance_test.go
package rpc

import (
	"encoding/json"
	"testing"
)

const (
	opaqueMCPHandle   = "opaque:mcp/01-do-not-parse"
	opaqueSkillHandle = "opaque:skill/02-do-not-parse"
)

func TestCatalogSearchResultPreservesCandidateSemantics(t *testing.T) {
	result, err := unmarshalCatalogSearchResult([]byte(`{
		"kind":"succeeded",
		"searchId":"search-01",
		"candidates":[
			{
				"kind":"mcp-server",
				"handle":"opaque:mcp/01-do-not-parse",
				"handleExpiresAt":"2026-09-02T12:00:00Z",
				"mediaType":"application/mcp-server-card+json",
				"installability":"installable",
				"displayName":"Example MCP",
				"rawCard":{"secret":"must-not-survive"},
				"source":{"kind":"url","url":"https://catalog.example/mcp.json"},
				"provenance":{
					"authority":"catalog.example",
					"observedAt":"2026-09-02T11:00:00Z",
					"mediaType":"application/mcp-server-card+json"
				}
			},
			{
				"kind":"ai-skill",
				"handle":"opaque:skill/02-do-not-parse",
				"handleExpiresAt":"2026-09-02T12:00:00Z",
				"mediaType":"application/ai-skill",
				"installability":"not-installable-kind",
				"displayName":"Example skill",
				"rawCard":{"secret":"must-not-survive"},
				"source":{"kind":"embedded"},
				"provenance":{
					"authority":"catalog.example",
					"observedAt":"2026-09-02T11:00:00Z",
					"mediaType":"application/ai-skill"
				}
			}
		],
		"truncated":false,
		"negotiated":{
			"runtimeProtocolVersion":1,
			"grantedCapabilities":["mcp-server-card","ai-skill-discovery"]
		}
	}`))
	if err != nil {
		t.Fatalf("unmarshal catalogue success: %v", err)
	}

	success, ok := result.(*CatalogSearchSucceeded)
	if !ok {
		t.Fatalf("catalogue result = %T, want *CatalogSearchSucceeded", result)
	}
	mcp, ok := success.Candidates[0].(*CatalogMCPServerCandidate)
	if !ok {
		t.Fatalf("first candidate = %T, want *CatalogMCPServerCandidate", success.Candidates[0])
	}
	skill, ok := success.Candidates[1].(*CatalogAiSkillCandidate)
	if !ok {
		t.Fatalf("second candidate = %T, want *CatalogAiSkillCandidate", success.Candidates[1])
	}
	if mcp.Handle != opaqueMCPHandle || skill.Handle != opaqueSkillHandle {
		t.Fatalf("opaque handles changed: %q, %q", mcp.Handle, skill.Handle)
	}
	if _, ok := mcp.Source.(*CatalogCandidateSourceURL); !ok {
		t.Fatalf("MCP source = %T, want *CatalogCandidateSourceURL", mcp.Source)
	}
	if _, ok := skill.Source.(*CatalogCandidateSourceEmbedded); !ok {
		t.Fatalf("skill source = %T, want *CatalogCandidateSourceEmbedded", skill.Source)
	}

	encoded, err := json.Marshal(success)
	if err != nil {
		t.Fatalf("marshal catalogue success: %v", err)
	}
	var wire map[string]any
	if err := json.Unmarshal(encoded, &wire); err != nil {
		t.Fatalf("decode catalogue wire result: %v", err)
	}
	for _, candidate := range wire["candidates"].([]any) {
		fields := candidate.(map[string]any)
		for _, forbidden := range []string{"card", "cardData", "rawCard"} {
			if _, exists := fields[forbidden]; exists {
				t.Fatalf("candidate leaked %q: %s", forbidden, encoded)
			}
		}
	}
}

func TestCatalogSearchResultPreservesRefusalsAndFailures(t *testing.T) {
	tests := []struct {
		name    string
		payload string
		assert  func(*testing.T, CatalogSearchResult)
	}{
		{
			name:    "authentication required",
			payload: `{"kind":"authentication-required","reason":"no-credential","message":"Sign in is required."}`,
			assert: func(t *testing.T, result CatalogSearchResult) {
				if _, ok := result.(*CatalogAuthenticationRequiredError); !ok {
					t.Fatalf("result = %T, want *CatalogAuthenticationRequiredError", result)
				}
			},
		},
		{
			name:    "network failure",
			payload: `{"kind":"network-failure","reason":"timeout","retryAfterSeconds":30,"message":"The catalogue timed out."}`,
			assert: func(t *testing.T, result CatalogSearchResult) {
				failure, ok := result.(*CatalogNetworkFailureError)
				if !ok {
					t.Fatalf("result = %T, want *CatalogNetworkFailureError", result)
				}
				if failure.RetryAfterSeconds == nil || *failure.RetryAfterSeconds != 30 {
					t.Fatalf("retryAfterSeconds = %v, want 30", failure.RetryAfterSeconds)
				}
			},
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			result, err := unmarshalCatalogSearchResult([]byte(test.payload))
			if err != nil {
				t.Fatalf("unmarshal catalogue result: %v", err)
			}
			test.assert(t, result)
		})
	}
}

func TestCatalogSearchResultRejectsUnknownCandidateKinds(t *testing.T) {
	_, err := unmarshalCatalogSearchResult([]byte(`{
		"kind":"succeeded",
		"searchId":"search-unknown",
		"candidates":[{
			"kind":"future-kind",
			"handle":"opaque:future/03-do-not-parse",
			"rawCard":{"secret":"must-not-survive"}
		}],
		"truncated":false,
		"negotiated":{"runtimeProtocolVersion":1,"grantedCapabilities":[]}
	}`))
	if err == nil {
		t.Fatal("unknown catalogue candidate kind with rawCard must be rejected")
	}
}
